libsupport/include/katana/ProgressTracer.h

```diff
@@ -0,0 +1,233 @@
+#include "katana/config.h"
+
+/// This tracer does not currently support thread-local tracers or concurrency controls
+/// Functions should only be used in a single-threaded context
+///
+/// Best Practices:
+///   If possible always avoid creating raw ProgressSpans from ProgressTracer
+///   If possible always use ProgressScopes to handle ProgressSpans
+///   Only use 1 ProgressTracer in an execution, so they should be created at entry points
+///   Use ProgressScope's RAII to handle early returns (i.e. due to errors)
+///   Raw ProgressSpans should be used for special scenarios like tracing asynchronous calls
+///
+/// Notes:
+///   SharedMemSys and DistMemSys will initialize the global ProgressTracer to a JsonTracer
+///     later it will be set to a No-op Tracer, on Fini or the destructor they will call
+///     Close on the global ProgressTracer
+///   This prevents GetProgressTracer() from returning nullptr and ensures Tracers are closed
+///
+///   ProgressScope's will only close their ProgressSpan when their ProgressSpan is the active span
+///   This means that if a user exclusively uses ProgressScopes,
+///   Then parent ProgressSpans will not be closed until their child ProgressSpan's close
+///   i.e.
+///   {
+///      auto scope1 = tracer->StartActiveSpan("1");
+///      auto scope2 = tracer->StartActiveSpan("2");
+///      if (err) { return; }
+///      scope2.Close();
+///      auto scope3 = tracer->StartActiveSpan("3");
+///   }
+///   Always results in scope2's ProgressSpan finishing before scope1's
+///   and scope3's ProgressSpan finishing before scope1's if there is no error
+
+namespace katana {
+
+// This is needed to properly handle being passed const char* values
+// Once we submodule OpenTracing this will be removed and their Value class used instead
+// TODO(Patrick)
+using variant_type = std::variant<
+    bool, int, int64_t, uint32_t, uint64_t, float, double, std::string,
+    const char*>;
+class Value : public variant_type {
+public:
+  Value(bool x) noexcept : variant_type(x) {}
+  Value(int x) noexcept : variant_type(static_cast<int64_t>(x)) {}
+  Value(int64_t x) noexcept : variant_type(x) {}
+  Value(uint32_t x) noexcept : variant_type(static_cast<uint64_t>(x)) {}
+  Value(uint64_t x) noexcept : variant_type(x) {}
+  Value(double x) noexcept : variant_type(x) {}
+  Value(std::string x) noexcept : variant_type(x) {}
+  Value(const char* s) noexcept : variant_type(std::string(s)) {}
+};
+using Tags = std::vector<std::pair<std::string, Value>>;
+
+class ProgressScope;
+class ProgressSpan;
+class ProgressContext;
+
+class KATANA_EXPORT ProgressTracer {
+  friend class ProgressSpan;
+
+  void SetSpan(std::shared_ptr<ProgressSpan> span);
+
+protected:
+  ProgressTracer(uint32_t host_id, uint32_t num_hosts)
+      : host_id_(host_id), num_hosts_(num_hosts) {}
+
+public:
+  virtual ~ProgressTracer() = default;
+
+  static std::shared_ptr<ProgressTracer> GetProgressTracer() { return tracer_; }
```

Comment:
Do we have to crate a new owner here? Could this be `ProgressTracer *` instead?

Reply:
To your above comment I still need to add a helper function to allow the distributed case and then I plan on adding this to import on top of the existing logging to make sure everyone agrees with it and serve as an example, then I will refactor the codebase and remove ProgressLogger
So this is still an experimental API and not supposed to be used by developers

That makes sense, but if we allow multiple tracers to be used by a program concurrently then it could be an issue
But we probably don't need that feature, and we could just say we don't allow it

Reply:
![image](https://user-images.githubusercontent.com/10097959/127376372-f32bee1f-47d4-40a5-b4be-93f49843189d.png)

Reply:
> That makes sense, but if we allow multiple tracers to be used by a program concurrently then it could be an issue
> But we probably don't need that feature, and we could just say we don't allow it

I don't understand what this has to do with copying the shared pointer? Or are you talking about something else?

Reply:
if the only shared_ptr copy of the tracer is the static variable, and we set the tracer to something else, then there would be no refs to it and it would be deleted right?

Reply:
Is having a single global pointer really the best way to implement multiple tracers? That seems confusing to me. If I'm setting the global tracer I would expect all users of the tracer to use the tracer I set. It seems like an anti-pattern to allow parts of the code to potentially hold a reference to an old tracer.

I would rather keep things simple.

Reply:
pushed a change switching it to a raw pointer, should be fine, is that what you had in mind?

Reply:
There are two points here:

First, returning or taking `std::share_ptr` should be looked at with some suspicion. If you mean to just provide access as long as the callee is live, `const T&`, `T*` and even `T&` for parameters is preferred:

All these forms,
```cpp
T& foo();
void Foo(const T&);
void Foo(T*);
```
signal to developers that `t` is borrowed and only needs to be live as long as the callee uses it.

When we write `std::shared_ptr` (or even `std::unique_ptr`) we are signaling to something about the management of the underlying resource. `std::shared_ptr` is particularly troublesome because it means you don't know who will use this resource and it may outlive the caller.

Second, if you want to implement a singleton, you can use this pattern:

```
class Foo {};

Foo& GetCurrentFoo();
void SetCurrentFoo(Foo); // or Foo&& 

static std::unique_ptr<Foo> foo;
```

If you really want, you can make `GetFoo` and `SetFoo` static methods on the class `Foo` instead of free functions, but the point here is (1) to not mix static members with static instances and (2) not allow callers to access the singleton except through `GetFoo` and `SetFoo`.

Reply:
is there an issue with the pyarrow dependency now or is that me?

Reply:
are the latest changes what you had in mind Donald?  I think that makes the most sense after talking to you and Tyler here